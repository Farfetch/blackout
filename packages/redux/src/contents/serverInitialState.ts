import { contentEntries } from '../entities/schemas/content';
import { generateContentHash } from './utils';
import { INITIAL_STATE_CONTENT } from './reducer';
import { normalize } from 'normalizr';
import get from 'lodash/get';
import merge from 'lodash/merge';
import type { ServerInitialState } from '../types/serverInitialState.types';

/**
 * Converts server data from searchContents to store state.
 *
 * @param page - Params from the `page` object injected by the server.
 *
 * @returns Initial state for the contents reducer.
 */
const serverInitialState: ServerInitialState = ({ model }) => {
  if (!get(model, 'searchContentRequests')) {
    return { contents: INITIAL_STATE_CONTENT };
  }

  const { searchContentRequests } = model;

  return searchContentRequests.reduce((acc, item) => {
    const {
      filters: { codes, contentTypeCode },
      searchResponse,
    } = item;

    const hash = generateContentHash({ codes, contentTypeCode });

    const { entities, result } = {
      ...normalize({ hash, ...searchResponse }, contentEntries),
    };

    return merge(acc, {
      entities,
      contents: {
        searchResults: {
          [hash]: {
            result,
            isLoading: false,
            error: null,
          },
        },
      },
    });
  }, {});
};

export default serverInitialState;
