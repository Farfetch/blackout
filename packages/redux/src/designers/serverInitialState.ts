import { generateDesignerResultHash } from './utils';
import { INITIAL_STATE } from './reducer';
import get from 'lodash/get';
import parse from 'url-parse';
import type { Model, StoreState } from '../types';

/**
 * Converts server data from designers to store state.
 *
 * @function serverInitialState
 *
 * @param {object} page - Params from the `page` object injected by the server.
 * @param {object} page.model - Page model.
 * @param {object} page.model.designers - Page model.
 * @param {string} page.model.slug - Page model.
 * @param {string} page.model.subfolder - Page model.
 *
 * @returns {object} Initial state for the designers reducer.
 */
const serverInitialState = ({
  model,
}: {
  model: Model;
}): {
  designers: StoreState['designers'];
} => {
  if (!get(model, 'designers')) {
    return { designers: INITIAL_STATE };
  }

  const { slug } = model;
  const { query } = parse(slug);
  const hash = generateDesignerResultHash(query);

  return {
    designers: {
      hash,
      error: {},
      isLoading: {
        [hash]: false,
      },
      result: {
        [hash]: model,
      },
    },
  };
};

export default serverInitialState;
