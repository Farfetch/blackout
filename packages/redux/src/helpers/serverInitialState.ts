import merge from 'lodash/merge';
import reduce from 'lodash/reduce';

const serverInitialState = (...reducers: any[]) =>
  reduce(
    reducers,
    (result: { entities?: any }, reducer) => ({
      ...result,
      ...reducer,
      entities: merge(result.entities, reducer.entities),
    }),
    {},
  );
export default serverInitialState;
