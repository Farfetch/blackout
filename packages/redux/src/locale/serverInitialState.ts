import { INITIAL_STATE_LOCALE } from './reducer';
import { normalize } from 'normalizr';
import country from '../entities/schemas/country';
import isEmpty from 'lodash/isEmpty';
import type { ServerInitialState } from '../types/serverInitialState.types';

/**
 * Converts server data for locale into store state.
 *
 * @param data - Params injected by the server.
 *
 * @returns - Initial state for the locale reducer.
 */

const localeServerInitialState: ServerInitialState = ({ model }) => {
  if (isEmpty(model)) {
    return { locale: INITIAL_STATE_LOCALE };
  }

  const {
    countryCode,
    countryId,
    cultureCode,
    currencyCode,
    currencyCultureCode,
    newsletterSubscriptionOptionDefault,
    requestSourceCountryCode,
    subfolder,
  } = model;
  // Normalize it
  const { entities } = normalize(
    {
      code: countryCode,
      cultures: [cultureCode],
      currencies: [
        {
          isoCode: currencyCode,
          cultureCode: currencyCultureCode,
        },
      ],
      newsletterSubscriptionOptionDefault,
      platformId: countryId,
      defaultCulture: cultureCode,
      defaultSubfolder: subfolder,
    },
    country,
  );

  return {
    locale: {
      countryCode,
      sourceCountryCode: requestSourceCountryCode,
    },
    entities,
  };
};

export default localeServerInitialState;
