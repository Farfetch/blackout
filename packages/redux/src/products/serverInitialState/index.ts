/**
 * Products serverInitialState.
 *
 * @module products/serverInitialState
 * @category Products
 * @subcategory ServerInitialState
 */
import listsServerInitialState from './lists';
import productsServerInitialState from './products';
import type { Model, StoreState } from '../../types';

export default ({
  model,
  options,
}: {
  model: Model;
  options?: { productImgQueryParam?: string };
}): {
  entities: StoreState['entities'];
  products: {
    attributes: StoreState['products']['attributes'];
    colorGrouping: StoreState['products']['colorGrouping'];
    details: StoreState['products']['details'];
    fittings: StoreState['products']['fittings'];
    lists: StoreState['products']['lists'];
    measurements: StoreState['products']['measurements'];
    sizeGuides: StoreState['products']['sizeGuides'];
    sizes: StoreState['products']['sizes'];
    variantsByMerchantsLocations: StoreState['products']['variantsByMerchantsLocations'];
  };
} => {
  const { entities: listsEntities, ...listsState } = listsServerInitialState({
    model,
    options,
  });
  const { entities: productsEntities, ...productsState } =
    productsServerInitialState({ model, options });

  return {
    entities: { ...listsEntities, ...productsEntities },
    products: {
      ...listsState,
      ...productsState,
    },
  };
};
