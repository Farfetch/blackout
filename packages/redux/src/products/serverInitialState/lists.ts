import { generateProductsListHash, getSlug } from '../utils';
import { INITIAL_STATE } from '../reducer/lists';
import { normalize } from 'normalizr';
import get from 'lodash/get';
import parse from 'url-parse';
import productsList from '../../entities/schemas/productsList';
import type { ListsServerInitialState } from './types';

/**
 * Converts server data for a products list (listing or sets) to store state.
 *
 * @param data - Params injected by the server.
 *
 * @returns Initial state for the products lists reducer.
 */
const serverInitialState: ListsServerInitialState = ({
  model,
  options: { productImgQueryParam } = {},
}) => {
  if (!get(model, 'products')) {
    return { lists: INITIAL_STATE };
  }

  // Only passing specifically the properties that we see fit,
  // instead of spreading everything and end up with unnecessary data that
  // will only increase the payload/store.
  const {
    breadCrumbs,
    config,
    didYouMean,
    facetGroups,
    facetsBaseUrl,
    filterSegments,
    gender,
    genderName,
    id,
    name,
    products,
    redirectInformation,
    searchTerm,
    slug,
  } = model;
  const { pathname, query } = parse(slug, true);
  // Remove CDN required `json=true` param from query which breaks our
  // selectors and causes SSR de-optimization
  delete query.json;

  const builtSlug = getSlug(pathname);
  const isSet = model?.pageType === 'set';
  const hash = generateProductsListHash(builtSlug, query, { isSet });

  // Normalize it
  const { entities } = normalize(
    {
      breadCrumbs,
      config,
      didYouMean,
      facetGroups,
      facetsBaseUrl,
      filterSegments,
      gender,
      genderName,
      hash,
      id,
      name,
      productImgQueryParam, // Send this to the entity's `adaptProductImages`
      products,
      redirectInformation,
      searchTerm,
      slug,
    },
    productsList,
  );

  return {
    lists: {
      error: {},
      hash,
      isHydrated: {
        [hash]: true,
      },
      isLoading: {
        [hash]: false,
      },
    },
    entities,
  };
};

export default serverInitialState;
