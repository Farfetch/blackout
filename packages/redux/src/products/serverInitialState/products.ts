import { INITIAL_STATE as DETAILS_INITIAL_STATE } from '../reducer/details';
import { normalize } from 'normalizr';
import { INITIAL_STATE as SIZES_INITIAL_STATE } from '../reducer/sizes';
import get from 'lodash/get';
import isEmpty from 'lodash/isEmpty';
import productSchema from '../../entities/schemas/product';
import type { Model, StoreState } from '../../types';

/**
 * Converts server data for details of a product to store state.
 *
 * @memberof module:products/reducer
 *
 * @param {object} data - Params injected by the server.
 * @param {object} data.model - Page model with product details data.
 * @param {object} data.options - General options for any modification.
 * @param {object} data.options.productImgQueryParam - Query parameter to be
 * appended to each product image URL.
 *
 * @returns {object} Initial state for the product details reducer.
 */
const serverInitialState = ({
  model,
  options: { productImgQueryParam } = {},
}: {
  model: Model;
  options?: { productImgQueryParam?: string };
}): {
  details: StoreState['products']['details'];
  entities?: StoreState['entities'];
  sizes: StoreState['products']['sizes'];
} => {
  // Check if a model object is of a product detail page (type === 'Product')
  // - if not, do nothing
  if (isEmpty(model) || get(model, 'dataLayer.general.type') !== 'Product') {
    return {
      details: DETAILS_INITIAL_STATE,
      sizes: SIZES_INITIAL_STATE,
    };
  }

  // Only passing specifically the properties that we see fit,
  // instead of spreading everything and end up with unnecessary data that
  // will only increase the payload/store.
  const {
    breadCrumbs,
    colorSet,
    colorSwatch,
    complementaryInformation,
    imageGroups,
    liveModel,
    price,
    productAttributes,
    productRef,
    productSize,
    recommendedSet,
    relatedSets,
    result: productData,
    scaleId,
    sizes,
    sizeSet,
    slug,
  } = model;
  const productInfo = {
    // Send this to the entity's `adaptProductImages`
    productImgQueryParam,
    // Data needed to support product information
    attributes: productAttributes,
    breadCrumbs,
    colorSet,
    colorSwatch,
    complementaryInformation,
    imageGroups,
    liveModel,
    price,
    productRef,
    productSize,
    recommendedSet,
    relatedSets,
    // @ts-expect-error scaleId is duplicated on product payload
    scaleId,
    sizes,
    sizeSet,
    slug,
    ...productData,
    // The model already brings a `measurements` entry,
    // which is different from the
    // '/api/products/{id}/variantsMeasurements' response.
    // Thus, we need to override it since it'll always be requested async.
    measurements: undefined,
  };
  const { result: id, entities } = normalize(productInfo, productSchema);

  return {
    details: {
      error: {},
      id,
      isHydrated: {
        [id]: true,
      },
      isLoading: {
        [id]: false,
      },
    },
    entities,
    sizes: {
      error: {},
      isLoading: {
        [id]: false,
      },
    },
  };
};

export default serverInitialState;
