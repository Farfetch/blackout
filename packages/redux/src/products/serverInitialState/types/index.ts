export * from './lists.types';
export * from './products.types';
