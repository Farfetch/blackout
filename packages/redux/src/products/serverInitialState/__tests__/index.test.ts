import {
  mockDetailsModel,
  mockProductsListModel,
} from 'tests/__fixtures__/products';
import serverInitialState from '..';
import type { Model } from '../../../types';

describe('serverInitialState()', () => {
  it('should initialize server state for a products list', () => {
    const slug =
      '/en-pt/shopping/woman?pageIndex=1&sort=price&sortDirection=asc';
    // @ts-expect-error A lot of properties would need to be added to make the value comply with the type which are irrelevant for the test
    const model = {
      ...mockProductsListModel,
      slug,
    } as Model;

    const state = serverInitialState({ model });

    expect(state).toMatchSnapshot();
  });

  it('should initialize server state for products', () => {
    const state = serverInitialState({
      // @ts-expect-error A lot of properties would need to be added to make the value comply with the type which are irrelevant for the test
      model: mockDetailsModel as Model,
    });

    expect(state).toMatchSnapshot();
  });
});
