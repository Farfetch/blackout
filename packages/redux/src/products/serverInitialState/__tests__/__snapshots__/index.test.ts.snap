// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serverInitialState() should initialize server state for a products list 1`] = `
Object {
  "entities": Object {
    "facets": Object {
      "categories_144307": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 13,
        "description": "Women",
        "groupsOn": 0,
        "id": "categories_144307",
        "parentId": "categories_0",
        "slug": "women",
        "url": "women",
        "value": 144307,
        "valueUpperBound": 0,
      },
      "categories_144331": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 2,
        "description": "Accessories",
        "groupsOn": 0,
        "id": "categories_144331",
        "parentId": "categories_144307",
        "slug": "women-accessories",
        "url": "women-accessories",
        "value": 144331,
        "valueUpperBound": 0,
      },
      "categories_144424": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 1,
        "description": "Special Accessories",
        "groupsOn": 0,
        "id": "categories_144424",
        "parentId": "categories_144418",
        "slug": "women-specials-special-accessories",
        "url": "women-specials-special-accessories",
        "value": 144424,
        "valueUpperBound": 0,
      },
      "colors_1": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 4,
        "description": "BLACK",
        "groupsOn": 0,
        "id": "colors_1",
        "parentId": "colors_0",
        "slug": null,
        "url": "?colors=1",
        "value": 1,
        "valueUpperBound": 0,
      },
      "pricetype_0": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 13,
        "description": "FullPrice",
        "groupsOn": 0,
        "id": "pricetype_0",
        "parentId": "pricetype_0",
        "slug": "fullprice",
        "url": "fullprice",
        "value": 0,
        "valueUpperBound": 0,
      },
      "sizes_22_144307": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 5,
        "description": "36",
        "groupsOn": 144307,
        "id": "sizes_22_144307",
        "parentId": "sizes_0",
        "slug": null,
        "url": "?sizes=22",
        "value": 22,
        "valueUpperBound": 0,
      },
    },
    "products": Object {
      "123": Object {
        "customAttributes": undefined,
        "groupedEntries": undefined,
        "id": 123,
        "images": undefined,
        "merchant": undefined,
        "price": Object {
          "discount": Object {
            "excludingTaxes": undefined,
            "includingTaxes": undefined,
            "rate": undefined,
          },
          "excludingTaxes": undefined,
          "formatted": Object {
            "includingTaxes": undefined,
            "includingTaxesWithoutDiscount": undefined,
          },
          "includingTaxes": undefined,
          "includingTaxesWithoutDiscount": undefined,
          "isFormatted": true,
          "priceType": undefined,
          "promotionType": undefined,
          "tags": undefined,
          "taxes": Object {
            "amount": undefined,
            "rate": undefined,
            "type": undefined,
          },
          "type": undefined,
        },
        "prices": undefined,
        "sizes": undefined,
        "tag": Object {
          "id": undefined,
          "name": undefined,
        },
        "variants": undefined,
      },
      "321": Object {
        "customAttributes": undefined,
        "groupedEntries": undefined,
        "id": 321,
        "images": undefined,
        "merchant": undefined,
        "price": Object {
          "discount": Object {
            "excludingTaxes": undefined,
            "includingTaxes": undefined,
            "rate": undefined,
          },
          "excludingTaxes": undefined,
          "formatted": Object {
            "includingTaxes": undefined,
            "includingTaxesWithoutDiscount": undefined,
          },
          "includingTaxes": undefined,
          "includingTaxesWithoutDiscount": undefined,
          "isFormatted": true,
          "priceType": undefined,
          "promotionType": undefined,
          "tags": undefined,
          "taxes": Object {
            "amount": undefined,
            "rate": undefined,
            "type": undefined,
          },
          "type": undefined,
        },
        "prices": undefined,
        "sizes": undefined,
        "tag": Object {
          "id": undefined,
          "name": undefined,
        },
        "variants": undefined,
      },
    },
    "productsLists": Object {
      "listing/woman?pageindex=1&sort=price&sortdirection=asc": Object {
        "breadCrumbs": Array [],
        "config": Object {
          "pageIndex": 1,
        },
        "didYouMean": undefined,
        "facetGroups": Array [
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 1,
            "description": "Categories",
            "dynamic": 0,
            "format": "hierarchical",
            "key": "categories",
            "order": 0,
            "type": 6,
            "values": Array [
              Array [
                "categories_144307",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 2,
            "description": "Categories",
            "dynamic": 0,
            "format": "hierarchical",
            "key": "categories",
            "order": 0,
            "type": 6,
            "values": Array [
              Array [
                "categories_144331",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 3,
            "description": "Categories",
            "dynamic": 0,
            "format": "hierarchical",
            "key": "categories",
            "order": 0,
            "type": 6,
            "values": Array [
              Array [
                "categories_144424",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 0,
            "description": "Colors",
            "dynamic": 0,
            "format": "multiple",
            "key": "colors",
            "order": 5,
            "type": 11,
            "values": Array [
              Array [
                "colors_1",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 0,
            "description": "PriceType",
            "dynamic": 0,
            "format": "multiple",
            "key": "pricetype",
            "order": 6,
            "type": 14,
            "values": Array [
              Array [
                "pricetype_0",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 0,
            "description": "Sizes",
            "dynamic": 0,
            "format": "multiple",
            "key": "sizes",
            "order": 4,
            "type": 9,
            "values": Array [
              Array [
                "sizes_22_144307",
              ],
            ],
          },
        ],
        "facetsBaseUrl": "/en-pt/shopping",
        "filterSegments": Array [
          Object {
            "deep": 1,
            "description": "Women",
            "facetId": "categories_144307",
            "fromQueryString": false,
            "gender": 0,
            "key": "categories",
            "negativeFilter": false,
            "order": 0,
            "parentId": 0,
            "slug": "women",
            "type": 6,
            "value": 144307,
            "valueUpperBound": 0,
          },
        ],
        "gender": 0,
        "genderName": "Woman",
        "hash": "listing/woman?pageindex=1&sort=price&sortdirection=asc",
        "id": 120198,
        "name": "New arrivals",
        "products": Object {
          "entries": Array [
            123,
            321,
          ],
          "totalItems": 193,
          "totalPages": 10,
        },
        "redirectInformation": null,
        "searchTerm": null,
        "slug": "/en-pt/shopping/woman?pageIndex=1&sort=price&sortDirection=asc",
      },
    },
  },
  "products": Object {
    "details": Object {
      "error": Object {},
      "isHydrated": Object {},
      "isLoading": Object {},
    },
    "lists": Object {
      "error": Object {},
      "hash": "listing/woman?pageindex=1&sort=price&sortdirection=asc",
      "isHydrated": Object {
        "listing/woman?pageindex=1&sort=price&sortdirection=asc": true,
      },
      "isLoading": Object {
        "listing/woman?pageindex=1&sort=price&sortdirection=asc": false,
      },
    },
    "sizes": Object {
      "error": Object {},
      "isLoading": Object {},
    },
  },
}
`;

exports[`serverInitialState() should initialize server state for products 1`] = `
Object {
  "entities": Object {
    "brands": Object {
      "1053": Object {
        "id": 1053,
      },
    },
    "categories": Object {
      "136301": Object {
        "id": 136301,
        "name": "Shoes",
      },
    },
    "products": Object {
      "11766695": Object {
        "attributes": Array [
          Object {
            "id": 7,
            "name": "Sleeve Length",
            "values": Array [
              Object {
                "id": 12,
                "value": "Longsleeved",
              },
            ],
          },
          Object {
            "id": 8,
            "name": "Neckline",
            "values": Array [
              Object {
                "id": 14,
                "value": "Round Neck",
              },
            ],
          },
        ],
        "brand": 1053,
        "breadCrumbs": Array [
          Object {
            "link": "shopping/woman",
            "parent": true,
            "slug": "woman",
            "text": "Woman",
          },
        ],
        "categories": Array [
          136301,
        ],
        "colorSet": Array [
          Object {
            "image": "",
            "name": "stadium-goods-spider-print-t-shirt",
            "productId": 14074682,
            "url": "/en-pt/shopping/-stadium-goods-spider-print-t-shirt-14074682",
          },
        ],
        "colorSwatch": null,
        "complementaryInformation": Array [
          Object {
            "title": "Designer",
            "value": "<p>Product designed by Chloé, originally from France.</p>",
          },
          Object {
            "title": "Care Information",
            "value": "<p>Wash the item in low temperatures.</p>",
          },
        ],
        "customAttributes": null,
        "description": "Luxury Italian",
        "gender": 0,
        "groupedEntries": undefined,
        "id": 11766695,
        "images": Array [
          Object {
            "order": 1,
            "size": "120",
            "sources": Object {
              "120": "https://cdn-images.farfetch-contents.com/services-unknown-x-east-stadium-goods-tri-band-hoodie_14074610_18494976_120.jpg",
            },
            "url": "https://cdn-images.farfetch-contents.com/services-unknown-x-east-stadium-goods-tri-band-hoodie_14074610_18494976_120.jpg",
          },
        ],
        "liveModel": Object {
          "id": 0,
        },
        "measurements": undefined,
        "merchant": undefined,
        "preferedMerchant": Object {
          "merchantId": 9359,
        },
        "price": Object {
          "discount": Object {
            "excludingTaxes": 100.8133,
            "includingTaxes": 124,
            "rate": 40,
          },
          "excludingTaxes": 151.22,
          "formatted": Object {
            "includingTaxes": "186,00 €",
            "includingTaxesWithoutDiscount": "310,00 €",
          },
          "includingTaxes": 186,
          "includingTaxesWithoutDiscount": 310,
          "isFormatted": true,
          "priceType": undefined,
          "promotionType": undefined,
          "tags": Array [
            "VAT",
          ],
          "taxes": Object {
            "amount": 34.78,
            "rate": 23,
            "type": "VAT",
          },
          "type": undefined,
        },
        "prices": undefined,
        "productRef": undefined,
        "productSize": undefined,
        "recommendedSet": 1050,
        "relatedSets": Array [],
        "scaleId": 117,
        "shortDescription": "Rockstud sling-back flats",
        "sizeSet": undefined,
        "sizes": Array [
          Object {
            "globalQuantity": 2,
            "id": 21,
            "isOneSize": false,
            "isOutOfStock": false,
            "name": "36",
            "scale": 206,
            "scaleAbbreviation": "IT",
            "scaleDescription": undefined,
            "stock": Array [
              Object {
                "barcodes": Array [
                  "2013323497172",
                ],
                "merchantId": 11554,
                "price": Object {
                  "formatted": Object {
                    "includingTaxes": "4 300 €",
                    "includingTaxesWithoutDiscount": "4 300 €",
                  },
                },
                "purchaseChannel": null,
                "quantity": 2,
              },
            ],
          },
        ],
        "slug": "rockstud-sling-back-flats-12854475",
        "tag": Object {
          "id": undefined,
          "name": undefined,
        },
        "variants": undefined,
      },
    },
  },
  "products": Object {
    "details": Object {
      "error": Object {},
      "isHydrated": Object {
        "11766695": true,
      },
      "isLoading": Object {
        "11766695": false,
      },
    },
    "lists": Object {
      "error": Object {},
      "hash": null,
      "isHydrated": Object {},
      "isLoading": Object {},
    },
    "sizes": Object {
      "error": Object {},
      "isLoading": Object {
        "11766695": false,
      },
    },
  },
}
`;
